### PR TITLE
[Perl] emulate Python zip() for Perl

### PR DESCRIPTION
**Pull request by tlby**
An alternate version of zip() that can be used a bit more like python's version, and runs a bit faster.
```perl
use strict;
use warnings;
use Benchmark;
use AI::MXNet 'mx';

my @l1 = (0 .. 64);
my @l2 = (0 .. 64);
my @l3 = (0 .. 64);
my @l4 = (0 .. 64);

sub orig_zip
{
    my ($sub, @arrays) = @_;
    my $len = @{ $arrays[0] };
    for (my $i = 0; $i < $len; $i++)
    {
        $sub->(map { $_->[$i] } @arrays);
    }
}

Benchmark::cmpthese(-4, {
    orig => sub {
        my @l;
        orig_zip(sub {
            push @l, $_[0] + $_[1] + $_[2] + $_[3];
        }, \@l1, \@l2, \@l3, \@l4);
    },
    repl => sub {
        my @l;
        AI::MXNet::zip(sub {
            push @l, $_[0] + $_[1] + $_[2] + $_[3];
        }, \@l1, \@l2, \@l3, \@l4);
    },
    pure => sub {
        my @l = map { $_->[0] + $_->[1] + $_->[2] + $_->[3] }
            AI::MXNet::zip(\@l1, \@l2, \@l3, \@l4);
    },
});
```
gives

|    |   Rate|orig|repl|pure|
|----|-------|----|----|----|
|orig|21964/s|  --|-32%|-37%|
|repl|32190/s| 47%|  --| -7%|
|pure|34657/s| 58%|  8%|  --|

### CODE REVIEW

**Comment by tlby** [comment]
@sergeykolychev, the benchmarks show a slight performance boost when using `zip()` without callbacks, and it would also be closer to the python code.  Should this patch also retool the uses of `zip()` to take advantage of this?

**Comment by sergeykolychev** [comment]
@gautamkmr could you please investigate this jenkins failure, missing license headers in files in the julia-package submodule was why this Jenkins failed.
This pull don't have a julia-package submodule and master doesn't (yet) either.

**Comment by tlby** [comment]
#8175 adds the julia-package work and had tests pending at the same time this failure occurred.

**Comment by sergeykolychev** [comment]
@tlby if you have time and desire please retool the current zip (there's also enumerate that uses zip ) usage get those 8% but that's substantial work. If you embark on this project please make sure that you test all examples in addition to the passing tests.

**Comment by tlby** [comment]
The second commit refactors most of the uses of `zip()` to avoid the callback form, though the handler still supports that call style.  Both tests and examples seem happy to me.

**Comment by sergeykolychev** [comment]
@tlby 
cool man 
did you run the GPU examples as well ?

File: perl-package/AI-MXNet/t/test_base.t [hunk 1]
```diff
@@ -0,0 +1,108 @@
+use strict;
+use warnings;
+use Test::More;
+use AI::MXNet qw(mx);
+use AI::MXNet::TestUtils qw(same reldiff GetMNIST_ubyte GetCifar10);
```
Comment (sergeykolychev) [inline perl-package/AI-MXNet/t/test_base.t]: this line is not needed

**Comment by sergeykolychev** [comment]
@tlby also please see if it makes sense to retool enumerate calls

**Comment by tlby** [comment]
I'd like to work out `enumerate()` in a different PR, since this one got kinda big with the refactor.

**Comment by sergeykolychev** [comment]
Ok

**Comment by tlby** [comment]
> did you run the GPU examples as well ?

I haven't yet, but I will.

**Comment by tlby** [comment]
The GPU examples are healthy.

**Pull request merged**